wgengine/tsdns/tsdns.go

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) 2020 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package tsdns provides a Resolver struct capable of resolving
+// domains on a Taiscale network.
+package tsdns
+
+import (
+	"encoding/binary"
+	"errors"
+	"strconv"
+
+	"github.com/tailscale/wireguard-go/device"
+	"tailscale.com/types/logger"
+	"tailscale.com/wgengine/packet"
+)
+
+const (
+	// MaxQuerySize is the maximal size of a Magic DNS query.
+	MaxQuerySize = 512
+	// MaxResponseSize is the maximal size of a Magic DNS response.
+	MaxResponseSize = 512
+)
+
+const (
+	ipOffset      = device.MessageTransportHeaderSize
+	dnsDataOffset = ipOffset + packet.UDPDataOffset
+
+	bufferSize = dnsDataOffset + MaxResponseSize
+)
+
+// Resolver is a
+type Resolver struct {
+	logf logger.Logf
+
+	// ip is the IP on which the resolver is listening (default 100.100.100.100).
+	ip packet.IP
+	// port is the port on which the resolver is listening (default 53).
```

Comment:
careful how you describe zero/default values in Go.

The Go assumption/mantra is that the zero values are useful. So usually a Go programmer will assume that when you say a default, you're talking about the meaning of the field being zero. To be more explicitly you can write:

  // Zero means to use port 53.

Or:

  // If unspecified, port 53 is used.

But that's not what this does. The default only comes from the `NewResolver` func.